File: app/src/main/java/com/daniel/mobilepauker2/model/pauker_native/SearchHit.java
package com.daniel.mobilepauker2.model.pauker_native;


/**
 * A search hit, when using the search function in Pauker.
 * @author Ronny.Standtke@gmx.net
 */
public class SearchHit {

    private final Card card;
    private final Card.Element cardSide;
    private final int cardSideIndex;

    /**
     * Creates a new instance of SearchHit
     * @param card          the card
     * @param cardSide      the card side
     * @param cardSideIndex the index of the card side
     */
    public SearchHit(Card card, Card.Element cardSide, int cardSideIndex) {
        this.card = card;
        this.cardSide = cardSide;
        this.cardSideIndex = cardSideIndex;
    }

    /**
     * returns the card
     * @return the card
     */
    public Card getCard() {
        return card;
    }

    /**
     * returns the card side
     * @return the card side
     */
    public Card.Element getCardSide() {
        return cardSide;
    }

    /**
     * returns the index of the card side
     * @return the index of the card side
     */
    public int getCardSideIndex() {
        return cardSideIndex;
    }

    @Override
    public boolean equals(Object object) {
        if (object instanceof SearchHit) {
            // class cast
            SearchHit otherSearchHit = (SearchHit) object;

            // compare cards
            if (!card.equals(otherSearchHit.getCard())) {
                return false;
            }

            // here cardSide should never be BOTH_SIDES!
            switch (cardSide) {
                case FRONT_SIDE:
                    if (otherSearchHit.getCardSide() ==
                            Card.Element.REVERSE_SIDE) {
                        return false;
                    }
                    break;
                case REVERSE_SIDE:
                    if (otherSearchHit.getCardSide() ==
                            Card.Element.FRONT_SIDE) {
                        return false;
                    }
                    break;
                default:

            }

            // compare "yellow" mark index
            return cardSideIndex == otherSearchHit.getCardSideIndex();
        }

        return false;
    }

    @Override
    public int hashCode() {
        int hash = 7;
        hash = 31 * hash + card.getId().hashCode();
        hash = 31 * hash + cardSide.ordinal();
        hash = 31 * hash + cardSideIndex;
        return hash;
    }
}

File: app/src/main/java/com/daniel/mobilepauker2/pauker_native/SearchHit.kt
package com.daniel.mobilepauker2.pauker_native

/**
 * A search hit, when using the search function in Pauker.
 * @author Ronny.Standtke@gmx.net
 */
class SearchHit(val card: Card, val cardSide: Card.Element, val cardSideIndex: Int) {

    override fun equals(other: Any?): Boolean {
        if (other is SearchHit) {
            if (card != other.card) {
                return false
            }
            when (cardSide) {
                Card.Element.FRONT_SIDE -> if (other.cardSide ==
                    Card.Element.REVERSE_SIDE
                ) {
                    return false
                }
                Card.Element.REVERSE_SIDE -> if (other.cardSide ==
                    Card.Element.FRONT_SIDE
                ) {
                    return false
                }
                else -> {
                }
            }
            // compare "yellow" mark index
            return cardSideIndex == other.cardSideIndex
        }
        return false
    }

    override fun hashCode(): Int {
        var hash = 7
        hash = 31 * hash + card.id.hashCode()
        hash = 31 * hash + cardSide.ordinal
        hash = 31 * hash + cardSideIndex
        return hash
    }

}